src/pycsou/opt/solver/lsmr.py

```diff
@@ -0,0 +1,332 @@
+            beta = normb.copy()
+        else:
+            x = x0.copy()
+            u -= self._A.apply(x)
+            beta = xp.linalg.norm(u)
+
+        if beta > 0:
+            u *= 1 / beta
+            v = self._A.T.apply(u)
+            alpha = xp.linalg.norm(v)
+        else:
+            v = xp.zeros_like(x)
+            alpha = 0
+
+        if alpha > 0:
+            v *= 1 / alpha
+
+        mst = self._mstate
+
+        normar = alpha * beta
+        if normar == 0:
+            mst["trivial"] = True
+            if b.ndim == 1:
+                mst["x"] = xp.zeros(n)
+            elif b.ndim == 2:
+                mst["x"] = xp.zeros((b.shape[0], n))
+            return
+
+        mst["x"], mst["u"], mst["v"], mst["alpha"], mst["trivial"] = x, u, v, alpha, False
+
+        # Initialize variables for 1st iteration:
+        mst["zetabar"], mst["alphabar"] = alpha * beta, alpha
+        mst["rho"] = mst["rhobar"] = mst["cbar"] = 1
+        mst["sbar"] = 0
+        mst["h"], mst["hbar"] = v.copy(), xp.zeros_like(x)
+
+        # Initialize variables for estimation of ||r||:
+        mst["betadd"] = beta
+        mst["rhodold"] = 1
+        mst["betad"] = mst["tautildeold"] = mst["thetatilde"] = mst["zeta"] = mst["d"] = 0
+
+        # Initialize variables for estimation of ||A|| and cond(A):
+        mst["normA2"] = alpha**2
+        mst["maxrbar"], mst["minrbar"] = 0, 1e100
+        mst["normA"], mst["condA"] = xp.sqrt(mst["normA2"]), 1
+        mst["normx"], mst["normr"], mst["normar"] = 0, beta, normar
+        self._normb = normb
+
+        # Initialize variables for testing stopping rules:
+        mst["test1"], mst["test2"], mst["test3"] = 1, alpha / beta, None
+
+    @staticmethod
+    def _sym_ortho(a, b, xp):
+        """
+        Stable implementation of Givens rotation.
+        """
+        if b == 0:
+            return xp.sign(a), 0, xp.abs(a)
+        elif a == 0:
+            return 0, xp.sign(b), xp.abs(b)
+        elif xp.abs(b) > xp.abs(a):
+            tau = a / b
+            s = xp.sign(b) / xp.sqrt(1 + tau**2)
+            c = s * tau
+            r = b / s
+        else:
+            tau = b / a
+            c = xp.sign(a) / xp.sqrt(1 + tau**2)
+            s = c * tau
+            r = a / c
+        return c, s, r
+
+    def m_step(self):
+
+        mst = self._mstate
+        if mst["trivial"]:
+            return
+        xp = pycu.get_array_module(mst["x"])
+
+        x, u, v, alpha = mst["x"], mst["u"], mst["v"], mst["alpha"]
+        zetabar, alphabar = mst["zetabar"], mst["alphabar"]
+        rho, rhobar, cbar = mst["rho"], mst["rhobar"], mst["cbar"]
+        sbar, h, hbar = mst["sbar"], mst["h"], mst["hbar"]
+        betadd, betad, tautildeold, thetatilde = mst["betadd"], mst["betad"], mst["tautildeold"], mst["thetatilde"]
+        rhodold, zeta, d = mst["rhodold"], mst["zeta"], mst["d"]
+        normA2, minrbar, maxrbar = mst["normA2"], mst["minrbar"], mst["maxrbar"]
+
+        self._itn += 1
```

Comment:
Would it be useful to use the solver iteration counter `self._astate["idx"]`?

Reply:
Fixed that.